source/kraken/maintenance_worker.cpp

```diff
@@ -485,29 +485,45 @@ void MaintenanceWorker::handle_rt_in_batch(const std::vector<AmqpClient::Envelop
                                                                              size_t max_nb,
-                                                                             size_t timeout_ms,
                                                                              bool no_ack) {
     assert(consume_tag != "");
     assert(max_nb);
-
     std::vector<AmqpClient::Envelope::ptr_t> envelopes;
-    envelopes.reserve(max_nb);
-    size_t consumed_nb = 0;
+
+    // count ready messages (prefetched messages are excluded from this count)
+    auto count_messages_consumers = declare_queue_with_counts(queue_name);
+    LOG4CPLUS_DEBUG(logger, "Nb of message(s) ready in queue " << queue_name << ": " << count_messages_consumers.first);
+    LOG4CPLUS_DEBUG(logger, "Nb of consumer(s) on queue" << queue_name << ": " << count_messages_consumers.second);
+    if (count_messages_consumers.second > 1) {
+        // not deadly but suspicious: RT would be partial for this kraken
+        LOG4CPLUS_ERROR(logger, "There is more than one consumer on RabbitMQ's RT queue " << queue_name);
+    }
+
+    auto message_prefetch_count = size_t(conf.broker_prefetch_message_count());
+    auto max_message_to_retrieve = std::min(size_t(count_messages_consumers.first) + message_prefetch_count, max_nb);
+    envelopes.reserve(max_message_to_retrieve);
+
     auto begin = pt::microsec_clock::universal_time();
 
-    auto retrieving_timeout = conf.total_retrieving_timeout();
-    while (consumed_nb < max_nb
-           && (pt::microsec_clock::universal_time() - begin).total_milliseconds() < retrieving_timeout) {
+    // If there are only prefetched messages (or none), use short "prefetch" timeouts
+    // Otherwise use long (classic) timeouts.
+    bool has_ready_messages = count_messages_consumers.first > 0;
+    auto single_timeout = has_ready_messages ? conf.broker_timeout() : conf.broker_prefetch_timeout();
+    auto total_retrieving_timeout =
+        has_ready_messages ? conf.broker_total_retrieving_timeout() : conf.broker_prefetch_total_retrieving_timeout();
+
+    while (envelopes.size() < max_message_to_retrieve
+           && (pt::microsec_clock::universal_time() - begin).total_milliseconds() < total_retrieving_timeout) {
         AmqpClient::Envelope::ptr_t envelope{};
 
-        /* !
-         * The emptiness is tested thanks to the timeout. We consider that the queue is empty when
-         * BasicConsumeMessage() timeout.
-         * */
-        bool queue_is_empty = !channel->BasicConsumeMessage(consume_tag, envelope, timeout_ms);
-        if (queue_is_empty) {
+        bool queue_is_empty_or_unreachable = !channel->BasicConsumeMessage(consume_tag, envelope, single_timeout);
+        if (queue_is_empty_or_unreachable) {
+            LOG4CPLUS_ERROR(logger,
+                            "Could not retrieve all the messages counted for RabbitMQ's RT queue " << queue_name);
+            // going on with what was retrieved so far
             break;
```

Comment:
An idea by @pbench (correct me if needed) to simplify code and params would be:
1. put a `continue` here instead of a `break`
2. only one total_timeout (i.e. 30s), single timeout may be hardcoded
3. remove count and just go with what was retrieved during those 30s

This would simplify code, the inconvenient is to have a strict latency even if there is only one message (and tests would need a little parameter adaptation for that reason).